rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,1794 @@
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use std::mem::transmute;
+
+use crate::core;
+use crate::dcerpc::parser;
+use crate::log::*;
+//use core::STREAM_TOSERVER;
+use nom::number::Endianness;
+
+use std::cmp;
+
+// Constant DCERPC UDP Header length
+pub const DCERPC_UDP_HDR_LEN: i32 = 80;
+// FIRST flag set on the packet
+pub const DCERPC_UUID_ENTRY_FLAG_FF: u16 = 0x0001;
+
+// Value to indicate first fragment
+pub const PFC_FIRST_FRAG: u8 = 0x01;
+// Value to indicate last fragment
+pub const PFC_LAST_FRAG: u8 = 0x02;
+// Cancel was pending at sender
+pub const PFC_PENDING_CANCEL: u8 = 0x04;
+pub const PFC_RESERVED_1: u8 = 0x08;
+// supports concurrent multiplexing of a single connection.
+pub const PFC_CONC_MPX: u8 = 0x10;
+// only meaningful on `fault' packet; if true, guaranteed
+//  call did not execute.
+pub const PFC_DID_NOT_EXECUTE: u8 = 0x20;
+// `maybe' call semantics requested
+pub const PFC_MAYBE: u8 = 0x40;
+// if true, a non-nil object UUID was specified in the handle, and
+//  is present in the optional object field. If false, the object field
+// is omitted.
+pub const PFC_OBJECT_UUID: u8 = 0x80;
+
+pub const REASON_NOT_SPECIFIED: u8 = 0;
+pub const TEMPORARY_CONGESTION: u8 = 1;
+pub const LOCAL_LIMIT_EXCEEDED: u8 = 2;
+pub const CALLED_PADDR_UNKNOWN: u8 = 3; /* not used */
+pub const PROTOCOL_VERSION_NOT_SUPPORTED: u8 = 4;
+pub const DEFAULT_CONTEXT_NOT_SUPPORTED: u8 = 5; /* not used */
+pub const USER_DATA_NOT_READABLE: u8 = 6; /* not used */
+pub const NO_PSAP_AVAILABLE: u8 = 7; /* not used */
+
+// DCERPC Header packet types
+pub const DCERPC_TYPE_REQUEST: u8 = 0;
+pub const DCERPC_TYPE_PING: u8 = 1;
+pub const DCERPC_TYPE_RESPONSE: u8 = 2;
+pub const DCERPC_TYPE_FAULT: u8 = 3;
+pub const DCERPC_TYPE_WORKING: u8 = 4;
+pub const DCERPC_TYPE_NOCALL: u8 = 5;
+pub const DCERPC_TYPE_REJECT: u8 = 6;
+pub const DCERPC_TYPE_ACK: u8 = 7;
+pub const DCERPC_TYPE_CL_CANCEL: u8 = 8;
+pub const DCERPC_TYPE_FACK: u8 = 9;
+pub const DCERPC_TYPE_CANCEL_ACK: u8 = 10;
+pub const DCERPC_TYPE_BIND: u8 = 11;
+pub const DCERPC_TYPE_BINDACK: u8 = 12;
+pub const DCERPC_TYPE_BINDNAK: u8 = 13;
+pub const DCERPC_TYPE_ALTER_CONTEXT: u8 = 14;
+pub const DCERPC_TYPE_ALTER_CONTEXT_RESP: u8 = 15;
+pub const DCERPC_TYPE_AUTH3: u8 = 16;
+pub const DCERPC_TYPE_SHUTDOWN: u8 = 17;
+pub const DCERPC_TYPE_CO_CANCEL: u8 = 18;
+pub const DCERPC_TYPE_ORPHANED: u8 = 19;
+pub const DCERPC_TYPE_RTS: u8 = 20;
```

Comment:
There's a duplicate constants error while compiling as these constants exist both in smb and dcerpc. I tried importing these into smb but it had some issues and I'm not sure if its the right way.
What would be the right way to share variables? A common file/dir or something?

Reply:
You can use them from smb by doing something like `use crate::smb::...` I think? (or the reverse, update smb to use this here)